File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int N = s.length();
        int res = 0, left = 0;
        //all uppercase letter, not array is enough
        int[] map = new int[26];
        for(int i = 0; i < N; i++) {
            //put in
            char c1 = s.charAt(i);
            map[c1 - 'A']++;
            //take out if condition not satisfied
            //if the maxfreq character 
            //******hard to understand
            while(i - left + 1 - findMax(map) > k) {
                char c2 = s.charAt(left);
                map[c2 - 'A']--;
                left++;
            }
            //calculate
            res = Math.max(res, i - left + 1);
        }
        return res;
    }
    public int findMax(int[] arr) {
        return Arrays.stream(arr).max().getAsInt();
    }
}

File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        l, n , res, freq = 0, len(s), 0, 0
        dic = defaultdict(int)
        for i in range(n):
            dic[s[i]] += 1
            freq = max(freq, dic[s[i]])
            while i - l + 1 - freq > k:
                c2 = s[l]
                dic[c2] -= 1
                l += 1
            res = max(res, i - l + 1)
        return res